llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp

```diff
@@ -525,26 +650,149 @@ void propagateAspectsThroughCG(Function *F, CallGraphTy &CG,
+
+  for (Function *Callee : It->second) {
+    identifyPathsContainingConditionalCallers(Callee, CG, CurrentPath, Paths);
+  }
+
+  CurrentPath.pop_back();
+}
+
+void propagateConditionExpressionsAndConditionalAspectsThroughCG(
+    const PathsContainingConditionalCallersTy &Paths,
+    FuncToConditionExprsAndConditionalAspectsTy &ConditionsAndAspectsMap,
+    FunctionToConditionExprsMapTy &ConditionsMap,
+    FunctionToAspectsMapTy &AspectsMap) {
+  auto Contains =
+      [](const ConditionExprsAndConditionalAspectsTy &ConditionsAndAspectsVec,
+         const std::pair<ConditionsSetTy, AspectsSetTy> &Search) {
+        return std::any_of(
+            ConditionsAndAspectsVec.begin(), ConditionsAndAspectsVec.end(),
+            [&Search](const auto ConditionAndAspectsPair) {
+              return (ConditionAndAspectsPair.first == Search.first) &&
+                     (ConditionAndAspectsPair.second == Search.second);
+            });
+      };
+
+  // TODO: need to optimize finding duplicates and combining pairs with the same
+  // Condition Expressions or Aspects
+  for (const auto &Path : Paths)
+    for (int E = Path.size() - 1, I = E; I >= 0; --I) {
+      if (I != E) {
+        const auto &CalleeAspects = AspectsMap[Path[I + 1]];
+        AspectsMap[Path[I]].insert(CalleeAspects.begin(), CalleeAspects.end());
+        if (ConditionsMap[Path[I]].empty())
+          ConditionsMap[Path[I]] = ConditionsMap[Path[I + 1]];
+      }
+      auto NewPair =
+          std::make_pair(ConditionsMap[Path[I]], AspectsMap[Path[I]]);
+      if (!NewPair.first.empty() && !NewPair.second.empty()) {
+        if (ConditionsAndAspectsMap[Path[I]].empty())
+          ConditionsAndAspectsMap[Path[I]].push_back(NewPair);
+        else if (!Contains(ConditionsAndAspectsMap[Path[I]], NewPair))
+          ConditionsAndAspectsMap[Path[I]].push_back(NewPair);
+      }
+      if (I != E)
+        for (const auto &ConditionAndAspectsPair :
+             ConditionsAndAspectsMap[Path[I + 1]])
+          if (!Contains(ConditionsAndAspectsMap[Path[I]],
+                        ConditionAndAspectsPair))
+            ConditionsAndAspectsMap[Path[I]].push_back(ConditionAndAspectsPair);
+    }
+}
+
+void propagateConditionExpressionsThroughCG(
+    Function *F, const CallGraphTy &CG,
+    FunctionToConditionExprsMapTy &ConditionsMap) {
+  // TODO: re-write using Paths from identifyPathsContainingConditionalCallers
+  // func
+  const auto It = CG.find(F);
+  if (It == CG.end())
+    return;
+
+  FunctionToConditionExprsMapTy LocalConditions;
+  for (Instruction &I : instructions(F)) {
+    const auto *CI = dyn_cast<CallInst>(&I);
+    if (!CI)
+      continue;
+
+    Function *CalledFunction = CI->getCalledFunction();
+    if (CI->isIndirectCall() || !CalledFunction)
+      continue;
+
+    if (!CalledFunction->hasFnAttribute("sycl-call-if-on-device-conditionally"))
+      continue;
+
+    // Start the loop with the 2nd argument (counting from 0) as 0 arg by
```

Comment:
Can a conditionally-called function return non-`void`? Because if so, then hardcoding 2 is incorrect here - in certain cases when structures are returned, they could be returned via `sret` argument, i.e. there will be an extra pointer argument which is actually a return value and it will be added at the beginning of the list.